In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
print("👽 Hello.")

👽 Hello.


In [2]:
import os

csv_file_path = os.getenv('HOME')+'/project/modulab2/trade.csv'
trade = pd.read_csv(csv_file_path) 
trade.head()

기간 국가명      수출건수        수출금액      수입건수       수입금액       무역수지  기타사항
0  2015년 01월  중국  116932.0  12083947.0  334522.0  8143271.0  3940676.0   NaN
1  2015년 01월  미국   65888.0   5561545.0  509564.0  3625062.0  1936484.0   NaN
2  2015년 01월  일본   54017.0   2251307.0   82480.0  3827247.0 -1575940.0   NaN
3  2015년 02월  중국   86228.0   9927642.0  209100.0  6980874.0  2946768.0   NaN
4  2015년 02월  미국   60225.0   5021264.0  428678.0  2998216.0  2023048.0   NaN

 - 중복된 데이터를 찾아 제거할 수 있고, 결측치(missing data)를 제거하거나 채워 넣을 수 있다.
 - 데이터를 정규화시킬 수 있다.
 - 이상치(outlier)를 찾고, 이를 처리할 수 있다.
 - 범주형 데이터를 원-핫 인코딩할 수 있다.
 - 연속적인 데이터를 구간으로 나눠 범주형 데이터로 변환할 수 있다.

In [3]:
print('전체 데이터 건수:', len(trade))

전체 데이터 건수: 199


In [4]:
print('컬럼별 결측치 개수')
len(trade) - trade.count()

컬럼별 결측치 개수


기간        0
국가명       0
수출건수      3
수출금액      4
수입건수      3
수입금액      3
무역수지      4
기타사항    199
dtype: int64

In [5]:
trade = trade.drop('기타사항', axis=1)
trade.head()

기간 국가명      수출건수        수출금액      수입건수       수입금액       무역수지
0  2015년 01월  중국  116932.0  12083947.0  334522.0  8143271.0  3940676.0
1  2015년 01월  미국   65888.0   5561545.0  509564.0  3625062.0  1936484.0
2  2015년 01월  일본   54017.0   2251307.0   82480.0  3827247.0 -1575940.0
3  2015년 02월  중국   86228.0   9927642.0  209100.0  6980874.0  2946768.0
4  2015년 02월  미국   60225.0   5021264.0  428678.0  2998216.0  2023048.0

In [6]:
trade.isnull()

기간    국가명   수출건수   수출금액   수입건수   수입금액   무역수지
0    False  False  False  False  False  False  False
1    False  False  False  False  False  False  False
2    False  False  False  False  False  False  False
3    False  False  False  False  False  False  False
4    False  False  False  False  False  False  False
..     ...    ...    ...    ...    ...    ...    ...
194  False  False  False  False  False  False  False
195  False  False  False  False  False  False  False
196  False  False   True   True   True   True   True
197  False  False   True   True   True   True   True
198  False  False   True   True   True   True   True

[199 rows x 7 columns]

In [7]:
trade.isnull().any(axis=1)

0      False
1      False
2      False
3      False
4      False
       ...  
194    False
195    False
196     True
197     True
198     True
Length: 199, dtype: bool

In [8]:
trade[trade.isnull().any(axis=1)]

기간 국가명      수출건수  수출금액       수입건수       수입금액  무역수지
191  2020년 04월  미국  105360.0   NaN  1141118.0  5038739.0   NaN
196  2020년 06월  중국       NaN   NaN        NaN        NaN   NaN
197  2020년 06월  미국       NaN   NaN        NaN        NaN   NaN
198  2020년 06월  일본       NaN   NaN        NaN        NaN   NaN

우선 '수출건수', '수출금액', '수입건수', '수입금액', '무역수지' 열이 모두 결측치인 index 196, 197, 198을 삭제

In [9]:
trade.dropna(how='all', subset=['수출건수', '수출금액', '수입건수', '수입금액', '무역수지'], inplace=True)
print("👽 It's okay, no biggie.")

👽 It's okay, no biggie.


DataFrame의 dropna는 결측치를 삭제해주는 메서드입니다.
 - subset 옵션으로 특정 컬럼들을 선택했습니다.
 - how 옵션으로 선택한 컬럼 전부가 결측치인 행을 삭제하겠다는 의미로 'all'을 선택합니다. ('any': 하나라도 결측치인 경우)
 - inplace 옵션으로 해당 DataFrame 내부에 바로 적용시켰습니다.

In [10]:
trade[trade.isnull().any(axis=1)]

기간 국가명      수출건수  수출금액       수입건수       수입금액  무역수지
191  2020년 04월  미국  105360.0   NaN  1141118.0  5038739.0   NaN

index 191과 같이 수치형 데이터를 보완할 방법은 많습니다.

1. 특정 값을 지정해줄 수 있습니다. 그러나 결측치가 많은 경우, 모두 같은 값으로 대체한다면 데이터의 분산이 실제보다 작아지는 문제가 생길 수 있습니다.

2. 평균, 중앙값 등으로 대체할 수 있습니다. 1번에서 특정 값으로 대체했을 때와 마찬가지로 결측치가 많은 경우 데이터의 분산이 실제보다 작아지는 문제가 발생할 수 있습니다.

3. 다른 데이터를 이용해 예측값으로 대체할 수 있습니다. 예를 들어 머신러닝 모델로 2020년 4월 미국의 예측값을 만들고, 이 값으로 결측치를 보완할 수 있습니다.

4. 시계열 특성을 가진 데이터의 경우 앞뒤 데이터를 통해 결측치를 대체할 수 있습니다. 예를 들어 기온을 측정하는 센서 데이터에서 결측치가 발생할 경우, 전후 데이터의 평균으로 보완할 수 있습니다.

In [11]:
trade.loc[[188, 191, 194]]

기간 국가명      수출건수       수출금액       수입건수       수입금액       무역수지
188  2020년 03월  미국   97117.0  7292838.0  1368345.0  5388338.0  1904500.0
191  2020년 04월  미국  105360.0        NaN  1141118.0  5038739.0        NaN
194  2020년 05월  미국  126598.0  4600726.0  1157163.0  4286873.0   313853.0

In [12]:
trade.loc[191, '수출금액'] = (trade.loc[188, '수출금액'] + trade.loc[194, '수출금액'] )/2
trade.loc[[191]]

기간 국가명      수출건수       수출금액       수입건수       수입금액  무역수지
191  2020년 04월  미국  105360.0  5946782.0  1141118.0  5038739.0   NaN

In [13]:
trade.loc[191, '무역수지'] = trade.loc[191, '수출금액'] - trade.loc[191, '수입금액'] 
trade.loc[[191]]

기간 국가명      수출건수       수출금액       수입건수       수입금액      무역수지
191  2020년 04월  미국  105360.0  5946782.0  1141118.0  5038739.0  908043.0